sympy/matrices/random.py

```diff
@@ -0,0 +1,2378 @@
+
+    Examples
+    ========
+    >>> from sympy import Matrix, I
+    >>> from sympy.abc import a, b
+    >>> from sympy.matrices.random import complex_to_real
+
+    >>> z = a + b * I
+    >>> A = Matrix([[z]])
+    >>> A
+    Matrix([[a + I*b]])
+
+    >>> complex_to_real(A)
+    Matrix([
+    [ re(a) - im(b), re(b) + im(a)],
+    [-re(b) - im(a), re(a) - im(b)]])
+
+    >>> A = Matrix([[z, 0],[1, I]])
+    >>> A
+    Matrix([
+    [a + I*b, 0],
+    [      1, I]])
+    >>> complex_to_real(A)
+    Matrix([
+    [ re(a) - im(b), re(b) + im(a),  0, 0],
+    [-re(b) - im(a), re(a) - im(b),  0, 0],
+    [             1,             0,  0, 1],
+    [             0,             1, -1, 0]])
+
+    Parameters
+    ==========
+    mat : Matrix
+        a complex matrix
+
+    """
+    dim = _shape(mat)[0]
+    mat = mat or _eye(dim)
+    obj = super_elementary_matrix(2 * dim)
+    for i in range(dim):
+        for j in range(dim):
+            z_value = mat[i, j]
+            a, b = _re(z_value), _im(z_value)
+            obj[2 * i, 2 * j] = a
+            obj[2 * i, 2 * j + 1] = b
+            obj[2 * i + 1, 2 * j] = -b
+            obj[2 * i + 1, 2 * j + 1] = a
+    return obj
+
+
+def regular_to_singular(mat, rank=None):
+    r""" build matrix with linear combination of matrix columns to meet rank
+
+    Explanation
+    ===========
+    Let $\mathbf{I}$ be the $n \times n$ identity matrix
+    and $\mathbf{D}$ an diagonal matrix with only zero and one entries
+    of rank $r$ (i.e. all entries but $n-r$ diagonal entries are zero).
+
+    If a full rank matrix $\mathbf{A}$ is given, then
+
+    .. math::
+
+        \mathbf{B}
+        = \mathbf{D} \cdot
+        (\mathbf{I} + \mathbf{A} \cdot (\mathbf{I}-\mathbf{D}))
+        = \mathbf{D} + \mathbf{D} \cdot \mathbf{A} \cdot(\mathbf{I}-\mathbf{D})
+
+    is of rank $r$, too.  Some for $\mathbf{A}\cdot\mathbf{B}$
+    which contains $r$ columns with $\mathbf{A}$.
+
+    If $\mathbf{A}$ is an upper triangular matrix $\mathbf{B}$
+    as well $\mathbf{A} \cdot \mathbf{B}$ is one, too.
+
+    Here, $\mathbf{A} \cdot \mathbf{B}$ is returned.
+    Note, if $n=r$ the matrix $\mathbf{B}$ is the identity.
+
+    Examples
+    ========
+    >>> from sympy import Matrix
+    >>> from sympy.matrices.random import regular_to_singular
+    >>> m = Matrix([[1, 2, 3],[4, 5, 6], [7, 8 ,0]])
+    >>> m.rank()
+    3
+    >>> n = regular_to_singular(m, 2)
+    >>> n
+    Matrix([
+    [1, 2, 15],
+    [4, 5, 42],
+    [7, 8, 69]])
+    >>> n.rank()
+    2
+
+    Parameters
+    ==========
+    mat : Matrix
+        a complex matrix
+    rank : integer
+        rank of matrix
+
+    """
+    if rank is None:
+        return mat
+    dim = _shape(mat)[0]
+    if rank == dim:
+        return mat
+    i = _eye(dim)
+    p = permutation(dim)
+    d = projection(dim, (0, rank))
+    d = _multiply(p.inv(), d, p)
+    md = _multiply(mat, d)
+    return md + _multiply(md, mat, i - d)
```

Comment:
Unless I misunderstand the `_multiply` function here is `Mul` (it's confusing to give it a different name) which is really just a roundabout way of using `*` i.e. the above can just be:
```python
d = p.inv() * d * p
md = mat * d
return md + md * mat * (i - d)
```

Reply:
Fine, removed all renaming of sympy function and names, e.g. use now `I` instead of `_i` and `Mul` or `*` instead of `_multiply`. Hope this makes the code less misunderstanding.